sphinx_immaterial/graphviz.py

```diff
@@ -0,0 +1,483 @@
+    node: sphinx.ext.graphviz.graphviz,
+    code: str,
+    options: dict,
+    prefix: str = "graphviz",
+    imgcls: str = None,
+    alt: Optional[str] = None,
+    filename: Optional[str] = None,
+) -> Tuple[str, str]:
+
+    theme_options = self.builder.config["html_theme_options"]
+    ttf_font_paths = external_resource_cache.get_ttf_font_paths(self.builder.app)
+    font = theme_options["font"]["text"]
+
+    ttf_font = ttf_font_paths[(font, "400")]
+
+    code = _replace_resolved_xrefs(node, code)
+
+    var_replacements: Dict[str, str] = {}
+    replacement_to_var: Dict[str, str] = {}
+
+    def replace_var(var_text: str) -> str:
+        replacement_color = var_replacements.setdefault(
+            var_text, "#%06x" % (0x123456 + len(var_replacements))
+        )
+        replacement_to_var.setdefault(replacement_color, var_text)
+        return replacement_color
+
+    def replace_var_in_code(m: re.Match) -> str:
+        var_text = m.group(1)
+        replacement_color = replace_var(var_text)
+        return f'"{replacement_color}"'
+
+    fontcolor = replace_var("var(--md-code-fg-color)")
+    fontsize = "12"
+
+    command_line_options = [
+        "-Ncolor=" + replace_var("var(--md-graphviz-node-fg-color)"),
+        "-Nstyle=solid,filled",
+        "-Nfillcolor=" + replace_var("var(--md-graphviz-node-bg-color)"),
+        "-Nfontcolor=" + fontcolor,
+        "-Nfontsize=" + fontsize,
+        "-Nfontname=" + ttf_font,
+        "-Ecolor=" + replace_var("var(--md-graphviz-edge-color)"),
+        "-Efontcolor=" + fontcolor,
+        "-Efontsize=" + fontsize,
+        "-Efontname=" + ttf_font,
+        "-Gbgcolor=transparent",
+        "-Gcolor=" + replace_var("var(--md-graphviz-node-fg-color)"),
+        "-Gfontcolor=" + fontcolor,
+        "-Gfontsize=" + fontsize,
+        "-Gfontname=" + ttf_font,
+    ]
+
+    code = re.sub(r'"((?:var|calc)\s*\(.*?\))"', replace_var_in_code, code)
+
+    graphviz_dot = options.get("graphviz_dot", self.builder.config.graphviz_dot)
+
+    dot_cmd = [graphviz_dot]
+    dot_cmd.extend(command_line_options)
+    dot_cmd.extend(self.builder.config.graphviz_dot_args)
+    dot_cmd.append("-Tsvg")
+
+    config_info = get_adjusted_graphviz_config(self.builder.app, graphviz_dot)
+    with tempfile.TemporaryDirectory() as tempdir:
+        env = os.environ.copy()
+        if config_info is not None:
+            orig_lib_dir = os.path.dirname(config_info.orig_config_path)
+            new_lib_dir = os.path.join(tempdir, "plugins")
+            pathlib.Path(
+                os.path.join(tempdir, os.path.basename(config_info.orig_config_path))
+            ).write_bytes(config_info.new_config)
+
+            env["GVBINDIR"] = tempdir
+            if os.name == "nt":
+                _winapi.CreateJunction(orig_lib_dir, new_lib_dir)  # type: ignore[attr-defined]
+            else:
+                os.symlink(orig_lib_dir, new_lib_dir)
+            cwd = os.path.dirname(config_info.orig_config_path)
```

Comment:
There's no reason (that I can see) to create a junction on Windows. My local tests show
```py
            orig_lib_path = pathlib.PurePath(config_info.orig_config_path)
            new_lib_dir = pathlib.Path(tempdir, "plugins")
            pathlib.Path(tempdir, orig_lib_path.name).write_bytes(
                config_info.new_config
            )

            env["GVBINDIR"] = tempdir
            new_lib_dir.symlink_to(orig_lib_path.parent, target_is_directory=True)
            cwd = str(orig_lib_path.parent)
```
working fine on windows (the `target_is_directory` param is ignored on other Posix systems).

I don't think this is the breaking point since it gets through this fine when the dot syntax is embedded. I'm just trying to isolate anything specific to Windows...

Reply:
I'm not sure if you're aware, but the dot exe is in the same dir as the plugin DLLs (by default that's `C:\Program Files\Graphviz\bin`). I say this because I noticed they're in separate locations on Linux (or at least for my WSL install of graphviz).

I don't know if that means anything though.

Reply:
I added the `-v` arg to dot exe calls ,and I'm seeing:
```
dot - graphviz version 5.0.0 (20220707.1540)
libdir = "abs_path\to\AppData\Local\Temp\tmp88gcajyy"
Activated plugin library: plugins\gvplugin_core.dll
Using render: svg:core
Using device: svg:svg:core
Activated plugin library: plugins\gvplugin_dot_layout.dll
Using layout: dot:dot_layout
The plugin configuration file:
        abs_path\to\AppData\Local\Temp\tmp88gcajyy\config6
                was successfully loaded.
    render      :  dot dot_json fig gdiplus json json0 map mp pic pov ps svg tk visio vml xdot xdot_json 
    layout      :  circo dot fdp neato nop nop1 nop2 osage patchwork sfdp twopi
    textlayout  :  textlayout
    device      :  bmp canon cmap cmapx cmapx_np dot dot_json emf emfplus eps fig gif gv imap imap_np ismap jpe jpeg jpg json json0 metafile mp pic plain plain-ext png pov ps ps2 svg tif tiff tk vdx vml xdot xdot1.2 xdot1.4 xdot_json
    loadimage   :  (lib) bmp eps gif jpe jpeg jpg png ps svg
Activated plugin library: plugins\gvplugin_neato_layout.dll
Using layout: twopi:neato_layout
fontname: unable to resolve "abs_path\to\docs\_build\html\_static\fonts\3254c528e2ab56454a9f22191035c5fe.ttf"
root = A max steps to root = 1
Rank separation = 0.000 1.000
overlap: none value 0 scaling 0.0000
Edge separation: add=1 (3.200000,3.200000)
Creating edges using line segments
Using render: svg:core
Using device: svg:svg:core
gvRenderJobs %3: 0.00 secs.
```
Notice it is still unable to resolve the fonts. The path output is correct, but I've shortened it here.
Unsuccessful attempts:
- I tried using posix style path separators
- I tried using `file:///` URI
- I set the `DOTFONTPATH` env var to `ttf_font`'s folder and only passing the font file's name (with and without suffix) to the dot exe args.

Reply:
Seems like maybe the libgd-based text layout still isn't being used --- perhaps the relevant plugin to disable on Windows is different?

Reply:
CI passes when using symlinks!

Reply:
Great that it works with symlinks --- still unclear what the issue with junctions was.  A bit surprising that there would be no error in that case, although _winapi being an undocumented module may lack proper error handling.

It appears that "developer mode" has to be enabled for unprivileged users to create symlinks: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinka?redirectedfrom=MSDN

Reply:
>  _winapi being an undocumented module may lack proper error handling.

It definetly doesn't have error checking. In searching for the module's docs, I came across a post from a python core dev that said [creating junctions] wasn't actually used by python std libs and it is only there for tests despite the fact that it doesn't check for things like target existence.

> It appears that "developer mode" has to be enabled for unprivileged users to create symlinks

I must have enabled it when installing Visual Sudio at some point. Not sure how this will impact end-users of this extension. Win8 required development mode for people that wanted to install apps without using the Windows store (win10 or newer has sideloading enabled by default because win8 was the worst _ever_). I'm not too worried about this because people developing docs with graphs are often using it to document an app that they support (thus developer mode should already be enabled).

Reply:
Should probably add a warning to the docs about symlinks on Windows. IDK about MacOS because Apple doesn't deserve my money.

Reply:
As a workaround (for this config stuff workaround), we could create 1 temp dir that has copies of the needed DLL files and avoid symlinks all together. Currently, it seems a little repetitive to create a temp dir and symlink to installed DLL for each use of the `graphviz` directive.

Despite what I said in https://github.com/jbms/sphinx-immaterial/pull/144#discussion_r950682756, the DLLs and config file seem to always be located in the same dir - not surprising given what the Graphviz docs say about the `GVBINDIR`  env var.